File: agent/error_log.py
"""Crashes, recorded where an operator can actually see them.

The worker's exception handling is careful and local: a dead board never fails a
sweep, a raising sender never takes down the run. That is right, and it has a
cost — the failure is printed into a container log that rotates, and nobody
reads a container log until someone complains. Sentry has been wired since the
beginning and has never been switched on, because switching it on needs an
account on someone else's service.

So: the same database everything else already uses. `capture()` is best-effort
and swallows its own errors by construction — an error logger that raises turns
a handled failure into an outage, which is precisely the opposite of the point.

Grouped, not appended. One exception inside a retry loop fires hundreds of times
a minute; a row per occurrence is a table nobody can read. Same fingerprint
bumps a count.
"""
from __future__ import annotations

import hashlib
import json
import os
import traceback

import db

# The top of a traceback is the part anyone reads, and a full trace of a deep
# framework call is measured in kilobytes per occurrence.
MAX_STACK = 4000
MAX_MESSAGE = 500

SOURCE = os.environ.get("GRINDLY_ERROR_SOURCE", "worker")


def fingerprint(source: str, kind: str, message: str) -> str:
    """What makes two crashes 'the same bug'.

    Deliberately excludes the stack: the same fault reached through two call
    paths is still one thing to fix, and including the trace would file every
    one separately and bury the page in duplicates.
    """
    raw = f"{source}\n{kind}\n{message}".encode("utf-8", "replace")
    return hashlib.sha1(raw).hexdigest()


def _clean(text: str, limit: int) -> str:
    text = (text or "").strip().replace("\x00", "")
    return text[:limit]


def capture(exc: BaseException | None = None, *, kind: str = "", message: str = "",
            context: dict | None = None, source: str = "") -> None:
    """Record one failure. Never raises, never blocks the caller."""
    try:
        source = source or SOURCE
        if exc is not None:
            kind = kind or type(exc).__name__
            message = message or str(exc)
            stack = "".join(
                traceback.format_exception(type(exc), exc, exc.__traceback__)
            )
        else:
            kind = kind or "unhandled"
            stack = ""

        message = _clean(message, MAX_MESSAGE) or kind
        stack = _clean(stack, MAX_STACK)

        # Credentials and OTPs pass through worker exceptions often enough that
        # this cannot be optional — a crash report is the last place a password
        # should end up, and this table is read by an operator, not the user it
        # belongs to.
        try:
            import redact

            message = redact.redact(message)
            stack = redact.redact(stack)
        except Exception:  # noqa: BLE001
            pass

        blob = ""
        if context:
            try:
                blob = _clean(json.dumps(context, default=str), 2000)
            except Exception:  # noqa: BLE001
                blob = ""

        fp = fingerprint(source, kind, message)
        with db.conn() as c:
            # One statement, so two workers crashing at the same instant cannot
            # race between a SELECT and an INSERT and lose a count.
            c.execute(
                """
                INSERT INTO error_events
                    (id, fingerprint, source, kind, message, stack, context,
                     count, first_seen_at, last_seen_at)
                VALUES (?, ?, ?, ?, ?, ?, ?, 1, now(), now())
                ON CONFLICT (fingerprint) DO UPDATE SET
                    count = error_events.count + 1,
                    last_seen_at = now(),
                    stack = EXCLUDED.stack,
                    context = EXCLUDED.context,
                    -- A bug that comes back was not fixed. Clearing this is what
                    -- makes "new since I last looked" mean something.
                    resolved_at = NULL
                """,
                (db.cuid(), fp, source, kind, message, stack, blob),
            )
            c.commit()
    except Exception as e:  # noqa: BLE001
        # The one place a bare print is the right answer: we are already in the
        # failure path, and raising here would replace a handled error with an
        # unhandled one.
        print(f"[error_log] could not record a failure: {type(e).__name__}: {e}")


def recent(limit: int = 50, include_resolved: bool = False) -> list[dict]:
    """Newest-first, for the admin surface and for asking the box directly."""
    try:
        where = "" if include_resolved else "WHERE resolved_at IS NULL"
        with db.conn() as c:
            rows = c.execute(
                f"""SELECT fingerprint, source, kind, message, count,
                           first_seen_at, last_seen_at
                    FROM error_events {where}
                    ORDER BY last_seen_at DESC LIMIT ?""",
                (limit,),
            ).fetchall()
        return [dict(r) for r in rows]
    except Exception as e:  # noqa: BLE001
        print(f"[error_log] could not read failures: {e}")
        return []

File: src/lib/errorLog.ts
import { createHash } from "crypto";
import { prisma } from "./prisma";

/**
 * Crashes, recorded where an operator can see them — the web half of
 * agent/error_log.py, writing to the same table with the same fingerprint rule
 * so a fault that happens on both sides groups as one row.
 *
 * Sentry has been wired since the beginning and never switched on, because
 * switching it on needs an account on someone else's service. This needs
 * nobody's signup and no DNS.
 *
 * Every function here is best-effort by construction: an error logger that
 * throws turns a handled failure into an outage.
 */

const MAX_STACK = 4000;
const MAX_MESSAGE = 500;

/** What makes two crashes "the same bug". Must match agent/error_log.fingerprint. */
export function fingerprint(source: string, kind: string, message: string): string {
  return createHash("sha1").update(`${source}\n${kind}\n${message}`).digest("hex");
}

function clean(text: string, limit: number): string {
  return (text || "").trim().replaceAll("\0", "").slice(0, limit);
}

export type CaptureInput = {
  source?: "web" | "browser";
  kind?: string;
  message: string;
  stack?: string;
  context?: Record<string, unknown>;
};

export async function captureError(input: CaptureInput): Promise<void> {
  try {
    const source = input.source ?? "web";
    const kind = clean(input.kind || "unhandled", 120);
    const message = clean(input.message, MAX_MESSAGE) || kind;
    const stack = clean(input.stack || "", MAX_STACK);
    let context = "";
    if (input.context) {
      try {
        context = clean(JSON.stringify(input.context), 2000);
      } catch {
        context = "";
      }
    }
    const fp = fingerprint(source, kind, message);

    await prisma.errorEvent.upsert({
      where: { fingerprint: fp },
      create: { fingerprint: fp, source, kind, message, stack, context },
      update: {
        count: { increment: 1 },
        lastSeenAt: new Date(),
        stack,
        context,
        // A bug that comes back was not fixed. Clearing this is what makes
        // "new since I last looked" mean anything.
        resolvedAt: null,
      },
    });
  } catch (e) {
    // Already in the failure path; throwing here would replace a handled error
    // with an unhandled one.
    console.error("[errorLog] could not record a failure:", e);
  }
}

/** Wrap a route handler so a throw is recorded before it becomes a 500. */
export async function recordingErrors<T>(
  where: string,
  run: () => Promise<T>,
): Promise<T> {
  try {
    return await run();
  } catch (e) {
    const err = e as Error;
    await captureError({
      kind: err?.name || "Error",
      message: err?.message || String(e),
      stack: err?.stack,
      context: { where },
    });
    throw e;
  }
}
